File: solutions/lua/run-length-encoding/1/run-length-encoding.lua
return {
        
  encode = function(s)

    if string.len(s) == 0 then return "" end

    local res = {}
    local previousLetter = string.byte(s, 1)
    local count = 1
            
    for i = 2, string.len(s) do
        local currentLetter = string.byte(s, i)
        if currentLetter ~= previousLetter then
            if count > 1 then
                table.insert(res, tostring(count))
            end
            table.insert(res, string.char(previousLetter))
            previousLetter = currentLetter
            count = 1
        else
            count = count + 1        
        end
    end
        
    if count > 1 then
        table.insert(res, tostring(count))
    end
    table.insert(res, string.char(previousLetter))

    return table.concat(res)
  end,

  decode = function(s)
            
    local res = {}
            
    for count, letter in string.gmatch(s, "(%d*)([%l%u ])") do
        if string.len(count) == 0 then
            table.insert(res, letter)
        else
            table.insert(res, string.rep(letter, tonumber(count)))
        end
    end
        
    return table.concat(res)
  end
}

File: solutions/go/run-length-encoding/1/run_length_encoding.go
package encode

import (
    "strings"
    "strconv"
    "regexp"
)

// CodedLetterRe represents the regular expression for a coded letter.
// It represents either 'nX' where n is an integer or 'X' where n is implicitely
// equal to 1.
var codedLetterRe = regexp.MustCompile(`((\d*)([\w ]))`)

// EncodeLetter encodes a single letter with a count n by outputing 'nX'.
// If n is 1, then the output is 'X' (1 is implicit).
func encodeLetter(b *strings.Builder, letter rune, count int) {
    if count == 0 {
        return
    }
    if count > 1 {
		b.WriteString(strconv.Itoa(count))
    }
    b.WriteRune(letter)    
}

// DecodeLetter decodes a letter by outputing it value n time.
func decodeLetter(b *strings.Builder, letter string, count int) {
	for i := 0; i < count; i++ {
        b.WriteString(letter)
    }    
}

// RunLengthEncode encodes a string using the Run Length encoding (see instructions
// for a description of Run Lengh encoding).
func RunLengthEncode(input string) string {
	var res strings.Builder
    var count int
    var currentLetter rune
    for _, letter := range input {
        if currentLetter != letter {
            encodeLetter(&res, currentLetter, count)
            count = 1
            currentLetter = letter
        } else {
            count++
        }
    }
    encodeLetter(&res, currentLetter, count)
    return res.String()
}
// RunLengthDecode decodes a string using the Run Length encoding (see instructions
// for a description of Run Lengh encoding).
func RunLengthDecode(input string) string {
    codedLetters := codedLetterRe.FindAllStringSubmatch(input, -1)
    if codedLetters == nil {
        return ""
    }
    var res strings.Builder
    for _, codedLetter := range codedLetters {
        count := 1
        // the number 1 in front of an encoded letter is omitted.
        if len(codedLetter[2]) > 0 {
        	// the definition of codedLetterRe ensures that codedLetter[2] is a string
            // of digits so we can assume the error returned by Atoi is always nil.
            count, _ = strconv.Atoi(codedLetter[2])
        }
        decodeLetter(&res, codedLetter[3], count)
    }
	return res.String()
}
